src/main/java/players/rhea/RHEAPlayer.java

```diff
@@ -0,0 +1,333 @@
+    private int fmCalls = 0;
+    private int copyCalls = 0;
+
+    public RHEAPlayer() {
+        this(System.currentTimeMillis());
+    }
+
+    public RHEAPlayer(RHEAParams params) {
+        randomGenerator = new Random(params.getRandomSeed());
+        this.params = params;
+        setName("rhea");
+    }
+
+    public RHEAPlayer(long seed) {
+        this(new RHEAParams(seed));
+    }
+
+    public RHEAPlayer(IStateHeuristic heuristic) {
+        this(System.currentTimeMillis());
+        this.heuristic = heuristic;
+    }
+
+    public RHEAPlayer(RHEAParams params, IStateHeuristic heuristic) {
+        this(params);
+        this.heuristic = heuristic;
+    }
+
+    public RHEAPlayer(long seed, IStateHeuristic heuristic) {
+        this(new RHEAParams(seed));
+        this.heuristic = heuristic;
+    }
+
+    @Override
+    public AbstractAction getAction(AbstractGameState stateObs, List<AbstractAction> actions) {
+        ElapsedCpuTimer timer = new ElapsedCpuTimer();  // New timer for this game tick
+        avgTimeTaken = 0;
+        acumTimeTaken = 0;
+        numIters = 0;
+        fmCalls = 0;
+        copyCalls = 0;
+
+        // Initialise individuals
+        population = new ArrayList<RHEAIndividual>();
+        for(int i = 0; i < params.populationSize; ++i)
+        {
+            population.add(new RHEAIndividual(params.horizon, params.discountFactor, getForwardModel(), stateObs, getPlayerID(), randomGenerator, heuristic));
+            fmCalls += population.get(i).length;
+        }
+
+        // Run evolution
+        boolean keepIterating = true;
+        while (keepIterating) {
+            runIteration(stateObs);
+
+            // Check budget depending on budget type
+            if (params.budgetType == PlayerConstants.BUDGET_TIME) {
+                long remaining = timer.remainingTimeMillis();
+                keepIterating = remaining > avgTimeTaken && remaining > params.breakMS;
+            } else if (params.budgetType == PlayerConstants.BUDGET_FM_CALLS) {
+                keepIterating = fmCalls < params.budget;
+            } else if (params.budgetType == PlayerConstants.BUDGET_COPY_CALLS) {
+                keepIterating = copyCalls < params.budget && numIters < params.budget;
+            } else if (params.budgetType == PlayerConstants.BUDGET_FMANDCOPY_CALLS) {
+                keepIterating = (fmCalls + copyCalls) < params.budget;
+            } else if (params.budgetType == PlayerConstants.BUDGET_ITERATIONS) {
+                keepIterating = numIters < params.budget;
+            }
+        }
+
+        // Return first action of best individual
+        return population.get(0).actions[0];
+    }
+
+    @Override
+    public RHEAPlayer copy() {
+        RHEAParams newParams = (RHEAParams) params.copy();
+        newParams.setRandomSeed(randomGenerator.nextInt());
+        return new RHEAPlayer(newParams);
+    }
+
+    private RHEAIndividual crossover(RHEAIndividual p1, RHEAIndividual p2)
```

Comment:
A general point on all the crossover methods. What happens if the action sequences are simply incompatible, and the action immediately after the crossover point is illegal?
For a simpe example, consider TicTacToe where we crossover two parents - the first one starts with playing in the middle square...while the second one starts in the top right corner, and plays in the middle second. This could lead to a genome that plays in the middle twice (which is illegal).